Claroty xdome integration

tomlandes opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues

## Description
This pull request adds a new pack named "Claroty xDome" and this Pack contains one new integration named "xDome".
In this integration an incident is a device-alert-pair and it's "Occurred" time is the device-alert-relation-update-time.

The Claroty xDome pack defines 7 xDome related Alert-types:

- xDome Custom Alert
- xDome Device Changes Alert
- xDome OT Activity Alert
- xDome Policy Deviation Alert
- xDome Risk Alert
- xDome Segmentation Alert
- xDome Threat Alert

And 38 xDome Incident-Fields:

- Claroty xDome Device First Seen
- Claroty xDome Device Likelihood Subscore Points
- Claroty xDome Device Type
- Claroty xDome Device Network
- Claroty xDome Device Known Vulnerabilities
- Claroty xDome Device Manufacturer
- Claroty xDome Device MAC
- Claroty xDome Device-Alert Detected Time
- Claroty xDome Device Likelihood Subscore
- Claroty xDome Alert Assignees
- Claroty xDome Alert ID
- Claroty xDome Device Labels
- Claroty xDome Device Risk Score
- Claroty xDome Device Impact Subscore
- Claroty xDome Device Effective Likelihood Subscore
- Claroty xDome Device Category
- Claroty xDome Alert Type
- Claroty xDome Device-Alert Updated Time
- Claroty xDome Alert Class
- Claroty xDome Alert Name
- Claroty xDome Device Known Vulnerabilities Points
- Claroty xDome Device Impact Subscore Points
- Claroty xDome Device Effective Likelihood Subscore Points
- Claroty xDome Device Purdue Level
- Claroty xDome Device Insecure Protocols Points
- Claroty xDome Alert Category
- Claroty xDome Device Last Seen
- Claroty xDome Device Insecure Protocols
- Claroty xDome Device Assignees
- Claroty xDome Alert Labels
- Claroty xDome Device-Alert Status
- Claroty xDome Device Name
- Claroty xDome Device Site Name
- Claroty xDome Device Internet Communication
- Claroty xDome Device UID
- Claroty xDome Device Subcategory
- Claroty xDome Device Risk Score Points
- Claroty xDome Device IP

The xDome integration also implements 3 command:

- **xdome-get-device-alert-relations** - fetch device-alert-relations data with custom filter & sort capabilities.
- **xdome-get-device-vulnerability-relations** - fetch device-vulnerability-relations data with custom filter & sort capabilities.
- **xdome-set-status-for-device-alert-relations** - resolve/unresolve device-alert-relations.

## Must have
- [ ] Tests
- [ ] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=34145) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=34145) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=34145) it.</sub>

ssokolovich commented on Packs/ClarotyXDome/Integrations/XDome/XDome_description.md:
Can we please add here more instructions on how to set up the instance? this is super helpful for the users. 
![image](https://github.com/demisto/content/assets/88268646/84e90cc9-2c4d-49ff-8ce1-9294e36cb9c8)

ssokolovich commented on Packs/ClarotyXDome/Classifiers/classifier-XDome.json:
```diff
@@ -0,0 +1,21 @@
+    "name": "xDome - Classifier",
+    "type": "classification",
+    "description": "Classifies xDome incidents.",
+    "keyTypeMap": {
```
Would like to suggest mapping default to `"Custom": "xDome Custom Alert"` (or something else that can be default),so in case you will add a new type (or the existing ones will be updated), a default classification will catch it :) 
Just a suggestion.
![image](https://github.com/demisto/content/assets/88268646/7378773b-f51a-4bb0-9d6e-0a7650c7a8f2)

ssokolovich reviewed (commented):
Hey @tomlandes! 
Thanks for your contribution to Claroty xDome. As for now, it looks good to me. Adding a couple of notes from my side:
1. Please add more details on how to set up the instance within the integration description.
2. Regarding the incident types, no playbooks/layouts are associated with them. Is that intentional? I can suggest setting up the default playbook or default layout for now, if that suits your needs.
3. Please avoid using "Extract All indicators from all fields" as this may significantly impact server performance. Instead, try to use extraction on specific fields that contain your information.
4. Usually, we suggest using our predefined OOTB fields. For example, instead of using `Claroty xDome Alert Name` you can map to the `Name` field (which also can be displayed the same info at the top of each create incident). Now, I looked into the created fields and I assume we can find others that can match -> but again up to you :).
5. Maybe consider adding a default classification as well. This might help you to maintain changes in the future.
6. Side note - didn't see any significant difference between the incidents in the mapper itself. If you are not planning to change it, maybe you can make the mapper easier to work with the default `common mapping` incident type ( which will map the same for all incidents regardless of their type).

Let me know your thoughts on my comments above. Here to assist and answer any questions. 
Cheers!

tomlandes commented:
> Hey @tomlandes! Thanks for your contribution to Claroty xDome. As for now, it looks good to me. Adding a couple of notes from my side:
> 
> 1. Please add more details on how to set up the instance within the integration description.
> 2. Regarding the incident types, no playbooks/layouts are associated with them. Is that intentional? I can suggest setting up the default playbook or default layout for now, if that suits your needs.
> 3. Please avoid using "Extract All indicators from all fields" as this may significantly impact server performance. Instead, try to use extraction on specific fields that contain your information.
> 4. Usually, we suggest using our predefined OOTB fields. For example, instead of using `Claroty xDome Alert Name` you can map to the `Name` field (which also can be displayed the same info at the top of each create incident). Now, I looked into the created fields and I assume we can find others that can match -> but again up to you :).
> 5. Maybe consider adding a default classification as well. This might help you to maintain changes in the future.
> 6. Side note - didn't see any significant difference between the incidents in the mapper itself. If you are not planning to change it, maybe you can make the mapper easier to work with the default `common mapping` incident type ( which will map the same for all incidents regardless of their type).
> 
> Let me know your thoughts on my comments above. Here to assist and answer any questions. Cheers!

Hey @ssokolovich :)
first of all - thank you for the super quick response!
regarding your notes:
1. Will do!
2. My original thought was that we will let our clients define playbooks that suit their needs (do you think it makes sense?). If you recommend to define a default playbook & layout I'll look into it.
3. I'm not sure I understand. do you mean that we should avoid mapping all the incident fields to custom fields?
4. I used the OOTB fields `Alert ID`, `Alert Name`, `Alert Category`, `Device ID`, `Device Name`. but I also mapped the same values to custom fields (e.x. `Claroty xDome Alert ID`) because I thought that it would be more convenient for our clients to just use our custom fields when they define a playbook without having to worry about whether or not the field is populated. But if this is a bad practice we can change that of course.
5. So I we decided to simplify things and just classify every incident as `Claroty xDome Alert`. You can see that in my recent commit I deleted all the other incident-types and added this general one (I saw other integrations doing a similar thing). is it OK?
6. the 5th note solves this as well.

Thanks!

ssokolovich commented:
> 1. Will do!
> 2. My original thought was that we will let our clients define playbooks that suit their needs (do you think it makes sense?). If you recommend to define a default playbook & layout I'll look into it.
> 3. I'm not sure I understand. do you mean that we should avoid mapping all the incident fields to custom fields?
> 4. I used the OOTB fields `Alert ID`, `Alert Name`, `Alert Category`, `Device ID`, `Device Name`. but I also mapped the same values to custom fields (e.x. `Claroty xDome Alert ID`) because I thought that it would be more convenient for our clients to just use our custom fields when they define a playbook without having to worry about whether or not the field is populated. But if this is a bad practice we can change that of course.
> 5. So I we decided to simplify things and just classify every incident as `Claroty xDome Alert`. You can see that in my recent commit I deleted all the other incident-types and added this general one (I saw other integrations doing a similar thing). is it OK?
> 6. the 5th note solves this as well.

Hey Tom,
Glad to hear that my feedback helped you!
2. That is also a good way to start, by providing the first content, feeling the customers' feedback, and then coming up with the adjustments and extra content such as Playbooks. The thing here is that all the mapping that you did won't be visible to the customers' analysts right away, as incident fields are presented in layouts (and not just context and mappers). Therefore, I suggest at least creating a Layout where they will be able to see your mapped information once they enter the XDOM incident. If you need more info we can jump on a call and I'll try to demonstrate what I mean.
3. Inside the Incident Type you've created, [the indicator extraction rules](https://docs-cortex.paloaltonetworks.com/r/Cortex-XSOAR/6.5/Cortex-XSOAR-Threat-Intel-Management-Guide/Indicator-Extraction-Modes) currently are will be performing extraction from all incident fields, which can significantly impact server performance. That's why I've suggested only extracting indicators from incident fields that contain the relevant data from the incident (instead of all).
4. No need for duplicated data, if you find an incident field that suits your needs, you don't need to create another one that will contain the same data. In case you didn't, only then you will need to create one. Hope that makes sense (again, better server performance).
5. Yes I can see that change. I'll cover everything again once all your changes have been submitted :) .
6. Correct, 5th note about going with 1 incident type works for that comment.
Let me know if you need anything else from my end.

maimorag reviewed (commented):
Hi @tomlandes,
Thank you for your contribution!
Good work :)

Please feel free to reach out to me with any questions - I'm available here or on slack :)
Thanks again

maimorag commented:
Hey @tomlandes,
please merge from master :)

maimorag commented:
Hey @tomlandes,
the integration display names: ['xDome API Base URL'] are invalid, Integration parameters display name should be capitalized and spaced using whitespaces and not underscores ( _ ).
Please fix this validation error.

maimorag commented:
Hey @tomlandes,
I'll be on PTO from May 12th to May 19th. Apologies for any delay in advance.

tomlandes commented:
Hi @maimorag,
Sorry for the delayed replay (unfortunately I'm experiencing some difficulties with upgrading to python3.10 in my XSOAR dev environment).
In the meantime I managed to push the fixes you requested :)
Let me know if there's anything else I can do on my end to advance the process.

tomlandes commented:
> If you need more info we can jump on a call and I'll try to demonstrate what I mean.

@ssokolovich I think me and my PM would like to take you on your offer for a short zoom meeting if that's ok.
We just want to fill-in some last gaps and see if we can wrap everything up.
Thanks!

ssokolovich commented:
> > If you need more info we can jump on a call and I'll try to demonstrate what I mean.
> 
> @ssokolovich I think me and my PM would like to take you on your offer for a short zoom meeting if that's ok. We just want to fill-in some last gaps and see if we can wrap everything up. Thanks!

Sounds good. The only thing I can suggest is Wednesday or Thursday around 5 pm Israel time next week. 
(We have upcoming holidays next week on Monday-Tuesday so that's why suggesting the following options).
Let me know what will suit you better and you can shoot me an invite ( you have my email from our last meeting).

tomlandes commented:
> > > If you need more info we can jump on a call and I'll try to demonstrate what I mean.
> > 
> > 
> > @ssokolovich I think me and my PM would like to take you on your offer for a short zoom meeting if that's ok. We just want to fill-in some last gaps and see if we can wrap everything up. Thanks!
> 
> Sounds good. The only thing I can suggest is Wednesday or Thursday around 5 pm Israel time next week. (We have upcoming holidays next week on Monday-Tuesday so that's why suggesting the following options). Let me know what will suit you better and you can shoot me an invite ( you have my email from our last meeting).

We are also available today if that works for you . if not then Wednesday will be fine as well :+1: 
We actually never had a meeting so it'll be good if you could share your email, mine is tom.la@claroty.com

ssokolovich commented:
> We are also available today if that works for you . if not then Wednesday will be fine as well 👍
> We actually never had a meeting so it'll be good if you could share your email, mine is [tom.la@claroty.com](mailto:tom.la@claroty.com)

Cool! sent an invite. let me know if the time slot suits you.
Cheers!

ssokolovich commented on Packs/ClarotyXDome/IncidentFields/incidentfield-Claroty_xDome_Device_Last_Seen.json:
```diff
@@ -0,0 +1,32 @@
+    "ownerOnly": false,
+    "description": "The date and time a device's NIC was last seen",
+    "cliName": "clarotyxdomedevicelastseen",
+    "type": "multiSelect",
```
is this should be multiselect?

tomlandes replied:
I chose to make it an `Array (Dynamic list)` because it looked like the most suitable option.
![WhatsApp Image 2024-05-16 at 12 55 25](https://github.com/demisto/content/assets/126682705/ca5b32d3-2c63-4231-854f-16d46b3e1cff)

Although you're right to think that it is a single-value field (it mostly is), sometimes devices will have multiple NICs and therefor this fields will contain multiple values (Last Seen for each NIC).
Do you think I should've defined something differently?

ssokolovich replied:
This depends on your values from the fetched JSON of the alert. If you are mapping multiple values use the multi-select option then :) just wanted to make sure that this is intentional.

ssokolovich commented on Packs/ClarotyXDome/IncidentFields/incidentfield-Claroty_xDome_Device_First_Seen.json:
```diff
@@ -0,0 +1,32 @@
+    "ownerOnly": false,
+    "description": "The date and time a device's NIC was first seen",
+    "cliName": "clarotyxdomedevicefirstseen",
+    "type": "multiSelect",
```
is this should be a multi-select?

ssokolovich replied:
This depends on your values from the fetched JSON of the alert. If you are mapping multiple values use the multi-select option then :) just wanted to make sure that this is intentional.

ssokolovich commented on Packs/ClarotyXDome/IncidentFields/incidentfield-Claroty_xDome_Device_Insecure_Protocols.json:
```diff
@@ -0,0 +1,32 @@
+    "ownerOnly": false,
+    "description": "The calculated level of the device’s ‘insecure protocols’ likelihood factor, such as \"Critical\", or \"High\"",
+    "cliName": "clarotyxdomedeviceinsecureprotocols",
+    "type": "shortText",
```
maybe this should be a multi-select?

tomlandes replied:
good point.
So I just committed changes to custom fields `Device Effective Likelihood Subscore`, `Device Impact Subscore`, `Device Insecure Protocols`, `Device Known Vulnerabilities`, `Device Likelihood Subscore`, `Device Risk Score`.
changed their type to `singleSelect` with `selectValues: ["Very Low", "Low", "Medium", "High", "Critical"]`

ssokolovich commented on Packs/ClarotyXDome/IncidentFields/incidentfield-Claroty_xDome_Device_Known_Vulnerabilities.json:
```diff
@@ -0,0 +1,32 @@
+    "ownerOnly": false,
+    "description": "The calculated level of the device’s ‘known vulnerabilities’ likelihood factor, such as \"Critical\", or \"High\"",
+    "cliName": "clarotyxdomedeviceknownvulnerabilities",
+    "type": "shortText",
```
maybe this one should be a multi-select?

ssokolovich commented on Packs/ClarotyXDome/Integrations/XDome/XDome.yml:
for some reason I don't see the following options:
* Fetch incident (yes or no)
* The default mapper and classifier

adding an image to emphasize 
![image](https://github.com/demisto/content/assets/88268646/a8b22b9f-f869-4d18-910f-3a7cadfe82dc)


Please double-check that you can see those from your end as well 🙏

tomlandes replied:
yes, this happens if you don't check the _Fetches incidents_ box in the _Integration Settings_.
But when I check it I can see those options in the _Instance Settings_
![WhatsApp Image 2024-05-16 at 11 43 59](https://github.com/demisto/content/assets/126682705/ac32778c-b1a1-4d5f-b62d-ee20466f25a6)

Should the _Fetches incidents_ box be checked by default?

ssokolovich replied:
Yes - otherwise all the default settings we put there for this mechanism will never be displayed to the user.
You can control in the instance default settings ( through the YML) if you want the fetch incident will be marked and enabled by default ( or not) - so this is up to you. 
But this check box marks if the mechanism will be visible to the users and if the entire integration supports this - so I assume this should be marked. 

[This is how it looks like when it is per instance we mark `do not fetch` by default](https://github.com/demisto/content/blob/09d357bfc07a05dd226b176d827889cc25b9d90b/Packs/MicrosoftExchangeOnPremise/Integrations/EWSv2/EWSv2.yml#L47)
![image](https://github.com/demisto/content/assets/88268646/43d9f099-a752-4862-8abd-5476e2009254)


[but the generic option of the integration, if it supports fetch incidents, looks like this](https://github.com/demisto/content/blob/09d357bfc07a05dd226b176d827889cc25b9d90b/Packs/MicrosoftExchangeOnPremise/Integrations/EWSv2/EWSv2.yml#L1107) - assuming that in your case this must be enabled.

tomlandes replied:
Thank you that was helpful. I added the `isfetch: true` to the yaml :+1:

ssokolovich reviewed (commented):
Adding some more notes from my end. Besides that, I think I'm ready for a demo from my end. @maimorag - waiting for your review as well :)

maimorag commented:
Hey @tomlandes,
the code looks good!
We're ready for a demo. Please check [this page](https://xsoar.pan.dev/docs/contributing/demo-prep), and let me know when you're available for one over DFIR. 
Feel free also to send me a recording of a demo.

tomlandes commented:
> Hey @tomlandes,
> the code looks good! We're ready for a demo. Please check [this page](https://xsoar.pan.dev/docs/contributing/demo-prep), and let me know when you're available for one over DFIR.
> Feel free also to send me a recording of a demo.

Great thank you @maimorag :)
Lets schedule a demo. Let me know when you're available (tom.la@claroty.com).
By the way, I wasn't able to obtain a "Developer Slack DFIR Username". could you help me with that (or refer me to someone)?

maimorag commented:
Hey @tomlandes,
I have sent you an email.
can you check again tomorrow morning if you can enter DFIR?
Thanks

maimorag commented:
Hey @tomlandes,
please merge from master.
Thanks :)
